### PR TITLE
Improve performance of latest endpoint in large data deployments

### PR DESCRIPTION
**Pull request by ctron**
## Summary by Sourcery

Optimize latest SBOM endpoint and related analysis by batching and caching graph queries, materializing describing CPE mappings, and tightening database indexing and migrations to scale better on large datasets.

New Features:
- Add materialized sbom_describing_cpe entity and population logic to speed up CPE lookups for describing components.
- Introduce configurable published-date cutoff for latest SBOM selection via environment variable.
- Provide CLI support for migrating the database up to a specific migration by name.
- Add container build artifacts (Containerfile and .containerignore) for building and running the trustd service.

Bug Fixes:
- Ensure ancestor walking and external SBOM resolution avoid cycles and handle multiple parents correctly while maintaining ranking semantics.
- Fix advisory sorting in tests by simplifying sort key usage to avoid inconsistent ordering.

Enhancements:
- Refactor latest-graph loading and CPE resolution to use batched queries, iterative DFS with caching, and reduced N+1 database calls.
- Add parallel graph loading with bounded concurrency and improved logging using truncated iter formatting.
- Replace inline describing CPE queries with reads from the new materialized table and adjust advisory product queries accordingly.
- Improve tracing and debug output for latest SBOM selection, cache hit/miss statistics, and ancestor graph traversal.
- Update time handling from chrono to time::OffsetDateTime in ranking logic and tests.

Build:
- Bump moka, petgraph, sea-orm, tokio, and uuid versions to newer releases.
- Add async-compression dependency for common utilities and unify uuid versions across crates.

Deployment:
- Introduce a multi-stage UBI-based container build for the trustd binary.

Documentation:
- Document the new TRUSTIFY_LATEST_SBOM_PUBLISHED_CUTOFF_DAYS environment variable in the env-vars reference.

Tests:
- Extend and adapt analysis, ingestor, and advisory tests to cover new ancestor traversal, CPE materialization, and external SBOM resolution behaviors.

Chores:
- Remove obsolete rust-toolchain and rustfmt configuration files and adjust recursion limits in several crates to accommodate deeper traversals.

### CODE REVIEW

**Comment by ctron** [comment]
@sourcery-ai review

File: modules/analysis/src/service/collector.rs
Comment (rh-jfuller) [inline modules/analysis/src/service/collector.rs:1]: while I was in this area I noticed that external SBOM traversal doesn't consume depth when it probably should (and can be a different PR).

File: modules/analysis/src/service/mod.rs
Comment (rh-jfuller) [inline modules/analysis/src/service/mod.rs:1]: another general observation while in this file: mpsc::unbounded_channel is unbounded ... we prob want to cap this off.

File: modules/analysis/src/service/load/rank.rs [hunk 1]
```diff
@@ -52,388 +54,827 @@ pub fn select() -> Select<sbom_node::Entity> {
...
-/// Assigns a rank to SBOMs within their specific CPE groups based on creation date which
-/// embodies the latest filter heuristics.
+// ─── Ranking ────────────────────────────────────────────────────────
+
+/// Assigns a rank to SBOMs within their specific CPE groups based on
+/// creation date which embodies the latest filter heuristics.
 ///
-/// This function simulates a SQL Window Function:
+/// Simulates a SQL Window Function:
 /// `DENSE_RANK() OVER (PARTITION BY cpe_id ORDER BY sbom_date DESC)`.
```
Comment (rh-jfuller) [inline modules/analysis/src/service/load/rank.rs:863]: comment should now be 'PARTITION BY (cpe_id, matched_name)'

File: modules/analysis/src/service/load/rank.rs [hunk 2]
```diff
@@ -52,388 +54,827 @@ pub fn select() -> Select<sbom_node::Entity> {
...
+        .collect::<HashSet<_>>()
+        .into_iter()
+        .collect();
+
+    let node_map: HashMap<_, _> = if all_node_ids.is_empty() {
+        HashMap::new()
+    } else {
+        sbom_node::Entity::find()
+            .filter(
+                Expr::col((sbom_node::Entity, sbom_node::Column::NodeId))
```
Comment (rh-jfuller) [inline modules/analysis/src/service/load/rank.rs:563]: not sure if we have to fix this now but if n+1 SBOMs share same node_id string (ex. "SPDXRef-DOCUMENT") we end up getting whatever was inserted last - which feels _uncontained_ ...

**Review by rh-jfuller — APPROVED** [review]
I think generally the PR is good  - I provided a few comments to discuss. 

We could add some more tests for rarer corner cases for **apply_rank** eg. I can think of a few things (mostly tie breaking scenarios) but none are blockers as all existing tests pass. There are also some things to consider from a DoS point of view (but that can be part of a larger analysis).

LGTM

File: modules/analysis/src/service/load/mod.rs [hunk 3]
```diff
@@ -723,10 +739,118 @@ fn q_columns() -> Columns {
...
+}
+
+impl Display for HitAndMiss {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        write!(
+            f,
+            "{} of {} ({:.1}%)",
+            self.hits,
+            self.total,
+            self.hits as f64 / self.total as f64 * 100.0
```
[the remark this answers is not in the document]
Comment (ctron) [reply modules/analysis/src/service/load/mod.rs:766]: That already seems to work.
